Add CitusDB to Travis

snopoke opened the pull request.

snopoke committed:
upgrade docker-compose

snopoke committed:
fix 'wait for services'

docker-compose 2.3 does not set the env vars that this used
to depend on. Using the service name will now
resolve to the IP

snopoke committed:
remove unused environent vars

As far as I can tell this is not used
by formplayer docker container and was
just copied from the web container

snopoke committed:
fix volume issue when running tests locally

snopoke committed:
pull all services in parallel

This is done by default now

snopoke committed:
add citusdb docker containers

snopoke committed:
add citusdb to docker settings and point icds-ucr to it

snopoke committed:
remove unused DB from docker localsettings

snopoke committed:
use reusedb / keepdb

Since the citus databaes is the default 'postgres' DB
it can't be deleted by Django either during setup
or teardown

snopoke committed:
make web depend on citus workers

snopoke committed:
comment on citus setup

snopoke committed:
fix wait for web in JS tests

millerdev commented on .travis.yml:
```diff
@@ -12,6 +12,7 @@ env:
     - TEST=python NOSE_DIVIDED_WE_RUN=bf JS_SETUP=yes
     - TEST=python-sharded-and-javascript JS_SETUP=yes TEST_MAKE_REQUIREMENTS=yes
   global:
+    - REUSE_DB=true  # can't drop the citus DB since it's the postgres database
```
Is the use of `postgres` a citus requirement or part of our setup? It seems bad to not have the ability to completely drop and re-create all database state when things go wrong, at least when testing locally. Is the `postgres` database used for application/test data, or is it just extension config stuff in there?

snopoke replied:
It would be great to find a way around this, especially for local testing. This is what Django would need to do:
1. Create the DB on the coordinator (easy)
2. Create the DB on the worker nodes
  - This would mean adding the DB configs to the settings file
3. Create the citus extension in all above DBs
  - Can be done with a DB migration
4. Add the worker nodes to the coordinator
  - Can also be done in a migration

So to do this in Django we'd need to have all the DB configs in the settings and we'd need 2 apps with migrations (one for the worker nodes and one for the coordinator) along with code in the DB router.

I started doing that but it's a lot of extra code etc. that's only necessary for tests. I'm open to ideas though.

millerdev replied:
> I started doing that but it's a lot of extra code etc. that's only necessary for tests.

So it sounds possible, but it's extra work and adds a bunch more code to maintain.

However, the alternative is that some database state will never be reset, even when running tests with `--reusedb=reset`. Also, when running tests locally in addition to a local HQ environment for manual testing, the tests and the local HQ environment will share the same citusdb database. Both of these seem important/bad enough that the extra work and code seems worth it to me.

Since it's for testing only, could the django apps and other code needed for this be added in `./testapps/`?

millerdev commented on corehq/tests/nose.py:
```diff
@@ -192,6 +192,8 @@ def __init__(self, tests, runner):
         self.skip_setup_for_reuse_db = reuse_db and reuse_db != "reset"
         self.skip_teardown_for_reuse_db = reuse_db and reuse_db != "teardown"
+        if reuse_db:
+            runner.keepdb = True
```
Curious if and why this became necessary now? Should we have been doing this all along?

snopoke replied:
I think we should have but it wasn't an option when this was first implemented. Setting this to True will prevent Django from trying to drop a DB if it already exists.

snopoke committed:
use correct DB alias

snopoke committed:
import

snopoke committed:
Merge branch 'master' into sk/citusdb-travis

snopoke committed:
citus-db setup via django migrations

snopoke committed:
Merge branch 'master' into sk/citusdb-travis

snopoke committed:
[ci skip] typo

snopoke committed:
Apply suggestions from code review

snopoke committed:
Merge pull request #23546 from dimagi/sk/citusdb-test-setup

citusdb test setup with Django

snopoke committed:
Merge branch 'master' into sk/citusdb-travis

snopoke commented:
not sure why this was failing or me since NickP seems to have got it working fine: https://github.com/dimagi/commcare-hq/pull/23431

snopoke commented:
wrong pr to close

snopoke committed:
change name to icds-ucr since other logic relies on it

https://github.com/dimagi/commcare-hq/commit/a49e381fef0668b687e0846eaddd35b5e6e6d75a

snopoke commented:
Still failing on ICDS export test :disappointed: :
```
custom.icds_reports.tests.test_export_data:TestExportData.test_awc_infrastructure_export ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_beneficiary_export ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_child_eight ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_child_five ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_child_four ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_child_nine ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_child_one ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_child_seven ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_child_six ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_child_ten ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_child_three ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_child_two ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_data_length ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_headers ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_children_export_info ... ok
custom.icds_reports.tests.test_export_data:TestExportData.test_demographics_export ... FAIL
======================================================================
FAIL: custom.icds_reports.tests.test_export_data:TestExportData.test_demographics_export
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/custom/icds_reports/tests/test_export_data.py", line 1760, in test_demographics_export
    'st1'
AssertionError: Lists differ: [[u'Demographics', [[u'State',... != [[u'Demographics', [[u'State',...
```

millerdev reviewed (commented):
from #23546
> This wasn't as bad as I'd thought. It does add 3-5min to each test run due to the 2 additional DBs that Django runs the migrations over.

That's an unfortunate addition to our test run time, considering that we've been hitting the travis limits again recently. I assume this only applies to the `python-sharded-and-javascript` test job, and not the others. Is that correct?

At some point I believe we will need to optimize the way django runs migrations to reduce the amount of time it takes during our test runs.

snopoke commented:
> That's an unfortunate addition to our test run time, considering that we've been hitting the travis limits again recently. I assume this only applies to the python-sharded-and-javascript test job, and not the others. Is that correct?

No this adds it to every run since the tests aren't isolated to just the sharded env. I don't know of an easy way of making an apps tests only run in one runner (as well as running all the `sql_backend` tests).

> At some point I believe we will need to optimize the way django runs migrations to reduce the amount of time it takes during our test runs.

100%! we have a lot of migrations and we should start squashing them. I tried this on the `form_processor` app once and it wasn't straight forward (since the migrations include a lot of `RunPython` and `RunSQL` which Django doesn't know how to squash).

snopoke commented on corehq/sql_db/routers.py:
```diff
@@ -25,6 +25,14 @@
 SYNCLOGS_APP = 'phone'
 AAA_APP = 'aaa'
 
+# this logic is necessary for test database setup
```
These vars in used in the router. Are you suggesting that I move the logic elsewhere and call it here?

snopoke replied:
moved to a separate router 5f142cfe2fc

snopoke committed:
add 'citus' command and compose file to docker setup

snopoke commented:
Build should get fixed by https://github.com/dimagi/commcare-hq/pull/23589

snopoke committed:
Merge branch 'master' into sk/citusdb-travis

snopoke committed:
move DB routing to new router

snopoke committed:
[ci skip] future imports

millerdev commented on testsettings.py:
```diff
@@ -3,6 +3,8 @@
 import settingshelper as helper
 from settings import *
 
+USING_CITUS = any(db for db in DATABASES.values() if db.get('ROLE') == 'citus_master')
```
nit: could be more succinctly phrased as
```py
any(db.get('ROLE') == 'citus_master' for db in DATABASES.values())
```

snopoke replied:
```suggestion
USING_CITUS = any(db.get('ROLE') == 'citus_master' for db in DATABASES.values())
```

snopoke committed:
rephrase USING_CITUS calculation
